authors/bruno-skvorc/polkadot/en_US/final.md

```diff
@@ -0,0 +1,186 @@
+
+RANDAO + VDF is secure by default, in that the number produced is guaranteed to be random.
+VRF is secure after the fact (suspected wrongdoing must be pointed out), but is significantly less resource intensive.
+
+## ICMP vs Crosslinks
+
+Crosslinks are how shards in Ethereum will be communicating. By telling the beacon chain that a state change has occured on the shard, the other shards which depend on fresh data from the changed shard need to query it manually. In other words, they need to know what they don't know so that they can request it. This will be a user-land implementation detail and will depend entirely on how wallets, IDEs, protocols and other users of Ethereum implement this communication.
+
+ICMP or inter-chain message passing is Polkadot's way of establishing communication across parachains and parathreads. Each chain and thread have an ingress and egress queue - a queue for incoming and outgoing messages, respectively. Messages are transactions meant for another chain or thread, which are forwarded to the destination parachain or parathread on every block through common collators - collators that are using both of the communicating chains. Since we assume that people will be running collator nodes for different chains at the same time, there should be a reasonably stable mesh effect going on allowing for messages to successfully gossip across to their destination. As a backup, in case some chains or threads are islands and have no other connections, validators will also make sure that ingress and egress queues of the chains and threads they are processing are empty. If not, then they will assume this role of message relayer.
+
+Both modes of communication - crosslinks and ICMP - are asynchronous and thus non-instant. The appearance of real-time communication will be a matter of userland implementation.
```

Comment:
I'm not sure what this means.  How could you have userland fake real-time communication from asynchronous messages?  Adding UI to show that it is "waiting"?

Reply:
Something like [this](https://github.com/ethereum/wiki/wiki/Sharding-FAQ#how-would-synchronous-cross-shard-messages-work) but also listening to receipts and pretending optimistically that a tx has happened as soon as it was initiated, so faking confirmations on the user level for trivial value amounts. [This](https://ethresear.ch/t/simple-synchronous-cross-shard-transaction-protocol/3097/7) is also interesting reading. This implementation is a nice PoC too: https://kauri.io/article/51c8d4c8908549468bb0f299f7c044b9/synchronous-cross-shard-transactions

Reply:
I dont know whether we can call them "asynchronous". What does asynchronous mean for you?